### PR TITLE
New payment card, switching apps, direct URLs, and more

### PR DESCRIPTION
**Pull request by kyranjamie**
- Removed Payment details notification in favour of changing button state
- Using `setState()` method currently from parent component
- Moved content/non-functional JSX to separate `content.js` file – I haven't seen this pattern  followed elsewhere, but find it helps to focus on behaviour when visual noise is removed
- Refactored `maker/index.js` to a functional component
- Create slice of state for maker portal in store

@hstove as discussed, will use `useReducer`, but will do separately as this requires moving `pages/maker/index.js` to a function component w/ hooks

### CODE REVIEW

File: stores/maker/reducer.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      }
+
+    case MakerActions.FETCH_APPS_DONE:
+      return {
+        ...state,
+        appIds: action.payload.apps.map(app => app.id),
+        appEntities: keyBy(action.payload.apps, 'id')
+      }
+
+    case MakerActions.SET_PAYMENT_DETAILS_COMPLETE:
```
[the remark this answers is not in the document]
Comment (kyranjamie) [reply stores/maker/reducer.js]: Totally agree, realised this myself as well. Can infer from a selector.

File: pages/maker/index.js [hunk 2]
```diff
@@ -2,22 +2,28 @@ import React, { useEffect } from 'react'
 import { Flex, Box, Type } from 'blockstack-ui'
 import { connect } from 'react-redux'
 
-import { selectMaker } from '@stores/maker/reducer'
-import { fetchApps } from '@stores/maker/actions'
+import { selectMaker, selectAppList, selectCurrentApp } from '@stores/maker/reducer'
```
Comment (hstove) [inline pages/maker/index.js]: Let's try and follow what we've been doing previously, and have all selectors inside `@stores/maker/selectors`

Comment (kyranjamie) [reply pages/maker/index.js]: I'll change this, and address the other comments tomorrow.

File: stores/maker/reducer.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import { keyBy } from 'lodash'
```
[the remark this answers is not in the document]
Comment (kyranjamie) [reply stores/maker/reducer.js]: I did 1, and just notice you're pushing toward number 2. 
Until app.co is at the level of maturity of having a test suite, wouldn't it better to use the lodash helper method, which itself has is rigorously tested, than copy/pasting in a helper method with no tests or documentation?

File: components/maker/payment/helpers.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+})
+
+export const validateSTX = memoize(addr => {
+  try {
+    c32Check.c32addressDecode(addr)
+    return true
+  } catch (error) {
+    return false
+  }
+})
```
Comment (aulneau) [inline components/maker/payment/helpers.js]: These exist within the package `@blockstack/stacks-utils` -> https://github.com/blockstack/stacks-utils/tree/feature/add-data-fetches-and-hardware-wallet-support

File: pages/maker/index.js [hunk 5]
```diff
@@ -1,98 +1,102 @@
...
-    loading: true,
-    apps: [],
-    errorMessage: null
-  }
-
-  componentDidMount() {
-    this.fetchApps()
-  }
+const MakerPortal = ({ query, params, apiServer, user, maker, errorMessage, appList, selectedApp, competionStatus, dispatch }) => {
+  console.log('Rendering component')
```
Comment (aulneau) [inline pages/maker/index.js]: This should be removed before merging

File: pages/maker/index.js [hunk 6]
```diff
@@ -1,98 +1,102 @@
...
+            />
+          </MakerStickyStatusBox>
+          <Box>
+            <MakerContentBox>
+              <Maker.PaymentDetails
+                app={app}
+                user={user}
+                apiServer={apiServer}
+                dispatch={dispatch}
+                // onPaymentDetailsComplete={() => dispatch(setPaymentDetailsComplete())}
```
Comment (aulneau) [inline pages/maker/index.js]: If we aren't using this it should be removed before merging

File: pages/maker/index.js [hunk 7]
```diff
@@ -1,98 +1,102 @@
...
+              />
+            </MakerContentBox>
+            <MakerContentBox>
+              <Maker.KYC
+                app={app}
+                user={user}
+                apiServer={apiServer}
+                // onKycComplete={() => dispatch(setKycComplete())}
```
Comment (aulneau) [inline pages/maker/index.js]: If we aren't using this it should be removed before merging

File: pages/maker/index.js [hunk 8]
```diff
@@ -1,98 +1,102 @@
...
+              />
+            </MakerContentBox>
+            <MakerContentBox>
+              <Maker.ParticipationAgreement
+                app={app}
+                user={user}
+                apiServer={apiServer}
+                // onLegalComplete={() => dispatch(setLegalComplete())}
```
Comment (aulneau) [inline pages/maker/index.js]: If we aren't using this it should be removed before merging

File: pages/submit.js [hunk 9]
```diff
@@ -258,54 +250,31 @@ class SubmitDapp extends React.Component {
   }
 
   render() {
+    // return <div>lkjsdflkjsdlfkjs</div>
```
Comment (aulneau) [inline pages/submit.js]: This should be removed

File: stores/maker/actions.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+export const SAVE_PAYMENT_DETAILS_FAIL = '[Maker Page] SAVE_PAYMENT_DETAILS_FAIL'
+
+export const SET_KYC_COMPLETE = '[Maker Page] SET_KYC_COMPLETE'
+export const SET_LEGAL_COMPLETE = '[Maker Page] SET_LEGAL_COMPLETE'
+
+export const FETCH_APPS = '[Maker Page] FETCH_APPS'
+export const FETCH_APPS_DONE = '[Maker Page] FETCH_APPS_DONE'
+export const FETCH_APPS_FAIL = '[Maker Page] FETCH_APPS_FAIL'
+
+export const SELECT_APP = '[Maker Page] SELECT_APP'
```
Comment (aulneau) [inline stores/maker/actions.js]: I've not seen this convention before, in the past typically actions would be of a value such as: `maker-page/SET_LOADING_DONE` -> https://github.com/blockstack/app.co/blob/3cd6b2cf6aaee68905988ef796c322315052ca7b/stores/mining/constants.js

I'm fine with any convention, I just think we should be consistent.

File: stores/maker/actions.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    })
+    await response.json()
+    dispatch(savePaymentDetailsDoneAction({ appId, btcAddress, stxAddress }))
+  } catch (error) {
+    dispatch(savePaymentDetailsFailAction(error))
+  }
+}
+
+export const fetchApps = ({ user, apiServer }) => async dispatch => {
+  dispatch(fetchAppsAction())
```
Comment (aulneau) [inline stores/maker/actions.js]: Should we dispatch this action after the user check and not before?

File: stores/maker/selectors.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+export const selectAppList = state => state.maker.appIds.map(id => state.maker.appEntities[id])
+
+export const selectCurrentApp = state => state.maker.appEntities[state.maker.selectedAppId]
+
+// export const selectCompetionStatus = state => {
+//   const selectedApp = state.appEntities[state.selectedAppId]
+//   return {
+//     paymentDetailsComplete
+//   }
+// }
```
Comment (aulneau) [inline stores/maker/selectors.js]: If we don't need this we should remove it before merging

Comment (kyranjamie) [reply stores/maker/selectors.js]: Sorry for the commented out stuff, I had some extra code to push that's stashed

File: components/maker/payment/helpers.js (hunk 4 above)
Comment (kyranjamie) [reply components/maker/payment/helpers.js]: Good to know, thanks.

**Comment by hstove** [comment]
Just gave this branch a look through. It seems like you pushed some half-finished code today, as the maker page doesn't build. It was just a few fixes to get it building, and I'm sure you just haven't pushed that yet. I also got the 'switching apps' functionality working, but it looks like it's not hydrating data yet. Nice work though, looking good overall!

Let's please get this branch in a working state by the demo tomorrow, so we can use the Heroku build for others to play with.

File: stores/maker/actions.js (hunk 11 above)
Comment (kyranjamie) [reply stores/maker/actions.js]: Done

File: pages/submit.js (hunk 9 above)
Comment (kyranjamie) [reply pages/submit.js]: Done

File: stores/maker/actions.js (hunk 10 above)
Comment (kyranjamie) [reply stores/maker/actions.js]: Done

File: components/maker/payment/helpers.js (hunk 4 above)
Comment (kyranjamie) [reply components/maker/payment/helpers.js]: Ran into issue using this lib, created issue here https://github.com/blockstack/stacks-utils/issues/4

**Pull request merged**